pkg/ordermanager/ordermanager.go

```diff
@@ -0,0 +1,411 @@
+/*
+Copyright 2020 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package ordermanager
+
+import (
+	context "context"
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	cryptorand "crypto/rand"
+	"crypto/tls"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"errors"
+	"fmt"
+	"strings"
+	"sync"
+	"time"
+
+	"golang.org/x/crypto/acme"
+	"golang.org/x/crypto/acme/autocert"
+	"golang.org/x/sync/errgroup"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"knative.dev/net-http01/pkg/challenger"
+	"knative.dev/pkg/apis"
+	logging "knative.dev/pkg/logging"
+)
+
+// Interface defines the interface for ordering new certificates.
+type Interface interface {
+	Order(ctx context.Context, domains []string, owner interface{}) (challenges []*apis.URL, cert *tls.Certificate, err error)
+}
+
+// OrderUpCallback is the signature of the function for notifying
+// owners that their order is up, and that they should invoke Order
+// again to pick it up.
+type OrderUpCallback func(owner interface{})
+
+const (
+	// Staging is the Let's Encrypt staging endpoint, which provides significantly
+	// higher quotas for use during development, but issues certificates that are
+	// not signed by it's root CA.
+	Staging = "https://acme-staging-v02.api.letsencrypt.org/directory"
+
+	// Production is the Let's Encrypt production endpoint, which issues certificates
+	// signed by their root CA.  This allows ~50 certificates to be issued per
+	// registered domain per week.
+	Production = autocert.DefaultACMEDirectory
+)
+
+var (
+	// Endpoint is the ACME API to use, it defaults to Production, but
+	// can be pointed at Staging or other compatible endpoints.
+	Endpoint = Production
+
+	// UserAgent is the HTTP user agent that is used with the ACME client,
+	// so that traffic from this client may be distinguished from others.
+	// It is intentionally exposed as a mutable variable in case users with
+	// to distinguish their traffic further.
+	UserAgent = "knative.dev/net-http01"
+)
+
+// New creates a new OrderManager.
+func New(ctx context.Context, cb OrderUpCallback, chlr challenger.Interface) (Interface, error) {
+	acctKey, err := ecdsa.GenerateKey(elliptic.P256(), cryptorand.Reader)
```

Comment:
we can also store the account key in a secret.
So when the controller is restarted, it does not register a new account which won't break the order that is being processed with the old account.